raft/util.go

```diff
@@ -70,6 +76,10 @@ func IsResponseMsg(msgt pb.MessageType) bool {
 	return isMsgInArray(msgt, isResponseMsg[:])
 }
 
+func IsLocalMsgTarget(id uint64) bool {
+	return id == LocalAppendThread || id == LocalApplyThread
```

Comment:
Is it necessary to introduce "threads" at the API level in this design, is it not deductible from the message type? I.e. I'm thinking just a `Local` target with no indication of threads. The receiver would then determine the "thread" based on the message metadata.

Also, currently / before this change, does `Ready` ever contain messages with `To == <this-node>`? If not, could we even avoid introducing the special target value(s), and instead send regular self-messages.

Reply:
Is it true that the requirement for the local `MsgStorageAppend` and `MsgStorageApply` messages is:
- When a `Resp` message is sent, we have the guarantee that the data in the request is persisted.
- App sends `MsgStorageAppendResp` messages back to raft in the same order `MsgStorageAppend` were requested.
- Ditto for `MsgStorageApply[Resp]` messages.

As discussed in person, this does not necessarily imply a requirement for running these requests sequentially at the app/storage side, which the "thread" semantics alludes to. For instance, storage could receive multiple appends (entries 11-20, and 21-30, and 31-40), parallelize them (if the storage engine can do so efficiently, a simple example is it has multiple disks and round robins entries). It may actually happen that latest entries are stored earlier.

Even for "apply" messages, where seemingly sequential semantics are necessary, some parallelization is possible, e.g. when batches are independent, most notably when the entire store is a CRDT.

As discussed, we probably don't want to deal with out-of-order replies at the raft side. One of the reasons: when we have an ack on a later batch, acks for earlier batches may never show up, so we would have to also add retry logic to raft. If it's possible to do out-of-order stuff, the storage implementation should do the necessary sync/reconciliation, and feed replies in the original order.

Reply:
Similarly to wanting to separate threads, I can also imagine sometimes wanting to **not** do so. Say, the storage is built so that it's more efficient to append entries and apply some earlier entries in the same batch/thread.

A silly example: maybe the "storage" is a single KV variable (or a few of them), and instead of separating the log and KV storages the implementation just piggybacks the applied KV data in every log entry. So, the handler would squash `MsgStorageAppend` and `MsgStorageApply` into a single sequence of entries and flush it; instead of flushing entries and the variable asynchronously / separately.

Reply:
Thanks for writing up these details from our discussion! 

> Is it true that the requirements for handling the local MsgStorageAppend and MsgStorageApply messages are:

Yes, the requirements that you laid out at all correct. The documentation that outlines these requirements is on `Config.AsyncStorageWrites`, and then again in `doc.go`. Please let me know if there are details missing there.

> Is it necessary to introduce "threads" at the API level in this design, is it not deductible from the message type? I.e. I'm thinking just a Local target with no indication of threads. The receiver would then determine the "thread" based on the message metadata.

I think where we landed during our discussion was that it was useful to assign a target to these messages for two reasons:
1. it allows a routing component to route to remote nodes, the local node, or local storage processes by looking solely at a message's `To` field. It avoids the need for this router to deconstruct the contents of messages in order to route them.
2. splitting the local storage targets into an "appender" and an "applier" makes it easier to discuss the ordering requirements for all messages to the same target, but absence of ordering requirements between messages to different targets.

There was also some discussion about whether the use of the term "thread" was too specific and restrictive to users of the library. This library leaves the storage and network implementation to the user, so there's a reasonable argument that the details of threading should also be left out of the API. My intention here was for "thread" to be thought of abstractly, more closely to an arbitrary "thread of execution" than a specific "pthread". In fact, if we wanted to refer to a specific threading implementation, we would probably want to call these "goroutines". There are parallels here with "storage" and "network", which the library evokes for clarity and intuition, but otherwise leaves the specific details to users.

> If not, could we even avoid introducing the special target value(s), and instead send regular self-messages?

We discussed the idea of an actor model and message passing briefly last week. In that analogy, it becomes important to differentiate a node ID from local storage actors. This is because the node ID is a reference to Raft state machine (this library). It would be ambiguous it the node ID could refer to multiple actors.

> As discussed in person, this does not necessarily imply a requirement for running these requests sequentially at the app/storage side, which the "thread" semantics alludes to.

> Similarly to wanting to separate threads, I can also imagine sometimes wanting to not do so.

These are helpful examples of how this library might be used to think through. Our goal when making this change should be to ensure that users of the library are provided with enough information to adhere to the requirements of the library (e.g. `MsgStorageAppend` msgs must be acked in order), but otherwise, given the flexibility to adhere to them however they want. For instance, the library refers to an "appender" and an "applier", but users are welcome to have the same goroutine play this role. Or to play this role directly in the state machine handling loop.

The library should then strive to place as few requirements on its use as possible, except where there is a meaningful benefit to restricting its use in a way that is not overly burdensome on users. The example we discussed was the reliable delivery of local storage messages. We could certainly allow for these messages to be dropped, but supporting this would add complexity and cost to the library. Since in-process communication is usually reliable (unlike across a network), it's hard to imagine cases where this restriction is particularly imposing to users of the library, so this seems like the right trade-off to make.